---
title: "EpiStatProfiler: a new R package for the qualitative analysis of DNA methylation"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{EpiStatProfiler: a new R package for the qualitative analysis of DNA methylation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
body {
text-align: justify}
</style>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width=8, 
  fig.height=6
)
```

### __19 Dec 2021__

## Background 
DNA methylation is an epigenetic modification - mainly occurring at CpG sites - involved in major biological mechanisms, such as genome stability, genome imprinting, and X-inactivation.
Association studies are typically aimed to identify genomic regions whose average DNA methylation differs among distinct conditions. However, analysing the methylation patterns of CpG sites at single-molecule level could provide additional insights about the cell-to-cell heterogeneity and the cell clonality within a sample. Epialleles (defined as all the methylation patterns observed in a given locus with __n__ CpG sites) information can be extracted from different types of bisulfite sequencing experiments. Here we present EpiStatProfiler, a new R package able to extract epiallele composition from both targeted and genome-wide bisulfite data. Moreover, EpiStatProfiler is provided with dedicated statistical tests aimed to compare the epiallele composition between two or more groups in order to identify genomic regions which are statistically different among conditions for their heterogeneity. 


```{r setup, warning= FALSE}
library(epistats)
```

# Workflow 

Here we describe the workflow of the analysis you can carry out using EpiStatProfiler. The functions that are provided by the tool can be classified into 4 categories: 

**a. Loading functions**

**b. Target regions design**

**c. Epialleles extraction**

**d. Statistics**


## **1. Data loading and filtering**
### 1.1 Data loading
Input data consist of the alignment file and a reference genome provided as BAM and FASTA files, rispectively. 
The paths of these files are to be specified as parameters in the loading function in order to be imported in the global environment. 
Input files are then loaded as GAlignment and DNAStringSet objects, rispectively.
You can get the single objects by subsetting them from the list returned by the __loadInput()__ function. 


```{r }
bamfile <- "/home/sarnataro/HD_RRBS/chr2_sorted.bam"
genomefile <- "/home/SHARED/Mus_musculus.GRCm38.dna_sm.primary_assembly.fa"
data <- loadInput(bamfile, genomefile)
alignments <- data[[1]]
## Visualize data objects
## Alignments - GAlignments object
knitr::kable(head(alignments, 5), "html")
```
```{r}
genome <- data[[2]]
## Visualize data objects 
### Genome - DNAStringSet object
genome
```

### 1.2 Filtering 
Filtering functions allow to subset the aligned reads covering only desired regions. To facilitate the demonstration, in this vignette we use a bam file contanining only reads aligned to the chromosome 2.
Anyway, by default, the function will select only standard chromosomes from the GAlignments object (excluding also the mitochondrial genome).

```{r}
alignments <- filterBAM(algn = alignments, 
                        keepStChr = TRUE, 
                        keepM = FALSE,
                        retainChr = NULL)
```


If you want to obtain epialleles information from a single chromosome (the chromosome 2 in the example), you can use the following function to drop seqlevels from the GAlignments object, as shown below. 

```{r}
alignments <- filterBAM(algn = alignments, retainChr = "2")
```

If you want to select aligned reads covering a set of specifc chromosomes, you need to provide them as a vector of characters when using the parameter __retainChr__.
Remember also to check the nomenclature used for the seqlevels in the analysis (you neeed to provide the right string).

```{r eval = FALSE}
alignments <- filterBAM(algn = alignments, 
                        keepStChr = TRUE, keepM = FALSE,
                        retainChr = c("chr1", "chr2", "chr3"))
## or 
alignments <- filterBAM(algn = alignments, 
                        keepStChr = TRUE, keepM = FALSE,
                        retainChr = c("1", "2", "3"))

```

## **2. Target region design**
### 2.1 Retrieveng Genomic Ranges that satisfy user-defined parameters. 

Genomic Ranges that satisfy coverage requirements are built from the GAlignments object. The user can specify the minimum coverage value used to perform the qualitative analysis. Here you decide the resolution of your analysis. All the genomic loci that will be analysed next, will be characterized by a minimum coverage set by the parameter __threshold__.
Genomic Ranges of a minimum size can be then selected (you can set this parameter once you have decided the method that will be used to generate target regions, as described in the section 2.2) 

```{r}
gregions <- epistats::filterByCoverage(alignments, 
                                       threshold = 50, 
                                       minsize = 8)
```

### 2.2 Building target regions 
EpiStatProfiler allows the users to design by their own the target regions that will be used to obtain their epiallele information. Moreover, you can build customized target regions using one of two different methods.
The figure below show the rationale that is behind these two methods. In the first case, regions are generated counting the CpG sites that fall in the given interval, while the other method consist of sliding windows of given lenght and step size. The two methods are explained in details in the next sections. 

![**Figure 1.**Regions design](/home/sarnataro/epistats/vignettes/TargetRegions.jpg){width=100%}

#### 2.2.1 Genomic Regions containing a fixed number of CpG sites 
You can build target regions having a different length size, but containing a fixed number of CpG sites (or any other pattern used to perform the analysis (e.g., CA", "CT")). The CpG sites can be specified throught the parameter __n__ and also the maximum length size of these regions can be set (this parameter should be set considering the average read length of your sequencing experiment).

```{r eval=FALSE}
regs <- makeBins(gr = gregions, 
                 Genome = genome, 
                 mode = "CG", 
                 n = 4, 
                 min.binsize = 8, 
                 max.binsize = 50, 
                 cores = 40)
```

This function returns a list of two GRanges objects. One containing the coordinated of the newly designed regions, and the other containing the coordinated of the CpG sites covered in the current analysis. 

```{r eval=FALSE}
## > to get the target regions 
regs <- regs$bins
```

#### 2.2.2 Genomic Regions from sliding windows of fixed length and step size 
You can decide to build genomic regions by using sliding windows of fixed length and step size. The user can set these values through function parameters, and also the minimum and the maximum number of CpG sites to be contained in the regios can be modified. 

```{r}
regs <- makeWindows(gr = gregions, 
                    Genome = genome, 
                    window = 50, 
                    step = 1, 
                    mode = "CG", 
                    min.C = 4, 
                    max.C = 10, 
                    cores = 40)
```

This function returns a list of two GRanges objects. One containing the coordinated of the newly designed regions, and the other containing the coordinated of the CpG sites covered by the latter. 

```{r}
## > to get the target regions 
regs <- regs$windows
```


## **3. Extracting epiallele composition**
We can obtain the epiallele composition information for each genomic interval given as input. A GRanges object containing all the intervals designed thorugh the previous steps is required. The user can adjust different parameters to perform a customized analysis. It can be decided to perform a stranded analysis or not. In this case, note that if you want to analyse non-CG methylation, only the stranded method is possible.
The check of bisulfite efficiency is also possible, and the user can decide to remove reads which show a low bisulfite efficiency from the analysis. The function also include an option to exclude reads with ambiguity (that eventually contain a polymorphism in that residue, etc...).
The function returns a list containing three different objects:
1. a compressed binary matrix containing the epiallele composition for each analysed genomic region. 
2. a summary dataframe containg the coordinates of the analysed regions and its relative metrics, such as mean CpG distance, average DNA methylation, etc...
3. a text file containing the coordinated of genomic intervals excluded from the analysis 

Before running the function, you need to specify the path of the folder where the output files will be saved thorough the 'pathDir' parameter. You also have to define the list of functions you want to operate on the binary epialleles matrices to obtain different metrics. EpiStatProfiler provides several functions such as shannon, epi, singleton, that will calculate metrics returned in the summary dataframe output. 
You can write the function you desire and then you can add it to the list passed in the __myfuns__ parameter. 

```{r}
## Set your analysis name
aname = "test_bam"
## Set the path to the folder where the output should be save
pathDir = "/home/sarnataro/HD_RRBS/"
## Decide the functions to use to calculate metrics
myfuns=list("dist"=cdist,
            "epi"=epi,
            "singleton"=singleton,
            "maxfreq"=maxfreq,
            "shannon"=shannon,
            "mean_met"=meanMeth,
            "num_cg"=numCG,
            "num_reads"=num_reads)

## Run epiAnalysis
out <- epiAnalysis(align = alignments,
                   bin = regs, 
                   aname = aname, 
                   threshold = 50, 
                   pathDir = pathDir, 
                   genome = genome,
                   myfuns = myfuns, 
                   bisu.Thresh = 0, 
                   stranded = FALSE,
                   mode = "CG", 
                   remove.Amb = TRUE, 
                   cores = 80, 
                   retain.reads = TRUE, 
                   get.cPos = FALSE)
```

1. The compressed binary matrix containing the epiallele composition for each analysed genomic region. 
```{r}
matrix <- out[[1]]
knitr::kable(head(matrix))
```

2. The summary dataframe containg the coordinates of the analysed regions and its relative metrics, such as mean CpG distance, average DNA methylation, etc...
```{r}
intervals <- out[[2]]
knitr::kable(head(intervals))
```

3. The text file containing the coordinated of genomic intervals excluded from the analysis 
```{r}
logout <- out[[3]]
knitr::kable(head(logout))
```

## **4. Statistics**
EpiStatProfiler is provided with a set of functions that perform statistycal test on the output obtained with the previous functions. Some of these functions work on the summary dataframes containing the metrics specified by the user, while others functions work on the compressed epialleles binary matrices.

For both types you need to build a dataframe containing your samples metadata as the one provided as example in this package. 
```{r}
metadata <- read.delim("/home/sarnataro/HD_RRBS/extdata/metadata.txt")
metadata
```


### 4.1 Functions that work on the summary dataframes 

#### 4.1.1 Finding regions that significantly differ for one metric using Wilcoxon test
```{r}
files <- c("/home/sarnataro/HD_RRBS/extdata/wt1.txt",
          "/home/sarnataro/HD_RRBS/extdata/wt2.txt",
          "/home/sarnataro/HD_RRBS/extdata/wt3.txt",
          "/home/sarnataro/HD_RRBS/extdata/HD1.txt",
          "/home/sarnataro/HD_RRBS/extdata/HD2.txt",
          "/home/sarnataro/HD_RRBS/extdata/HD3.txt")

## load data 
library(magrittr)
library(purrr)
summary <- files %>% map(read.table, header = TRUE, quote = '"')
names(summary) = metadata$Samples

## RUN diffStat 
diff <- diffStat(intervals_list = summary,
                 metadata = metadata,
                 colsamples = "Samples",
                 statistic = "shanNorm",
                 groupcol = "Group",
                 cores = 40,
                 reduce = FALSE)
```


### 4.2 Functions that work on compressed epialleles matrices 

```{r}
matrices <- c("/home/sarnataro/HD_RRBS/extdata/wt1mat.txt",
              "/home/sarnataro/HD_RRBS/extdata/wt2mat.txt",
              "/home/sarnataro/HD_RRBS/extdata/wt3mat.txt",
              "/home/sarnataro/HD_RRBS/extdata/HD1mat.txt",
              "/home/sarnataro/HD_RRBS/extdata/HD2mat.txt",
              "/home/sarnataro/HD_RRBS/extdata/HD3mat.txt")

## load data 
library(magrittr)
library(purrr)
epimatrices <- matrices %>% map(read.table, header = TRUE, quote = '"')
names(epimatrices) = metadata$Samples

## RUN epiStat
outstat <- epiStat(sample_list = epimatrices,
                   metadata = metadata,
                   colgroups = "Group",
                   colsamples = "Samples",
                   rmUnmeth = FALSE,
                   cores = 30,
                   minGroups = 2,
                   minSampleSize = 3,
                   reduce = TRUE)
```

### 4.3 Plotting functions

#### CCA 
```{r}
runCCA(samples = epimatrices, region = "1_136131584_136131653", metadata = metadata, printData = FALSE, rmUnmeth = FALSE)
```


```{r}
runRDA(samples = epimatrices, 
       region = "1_136131584_136131653",
       printData = FALSE, 
       rmUnmeth = FALSE)
```


